libavcodec/libdavs2.c

```diff
@@ -0,0 +1,206 @@
+
+    int decoded_frames;
+
+    davs2_picture_t  out_frame;  // output data, frame data
+    davs2_seq_info_t headerset;  // output data, sequence header
+
+}DAVS2Context;
+
+static av_cold ff_davs2_init(AVCodecContext *avctx)
+{
+    DAVS2Context *cad = avctx->priv_data;
+
+    /* init the decoder */
+    cad->param.threads      = 0;
+    cad->param.i_info_level = 0;
+    cad->decoder = davs2_decoder_open(&cad->param);
+    avctx->flags |= AV_CODEC_FLAG_TRUNCATED;
+
+    av_log(avctx, AV_LOG_VERBOSE, "decoder created. 0x%p\n", cad->decoder);
+    return 0;
+}
+
+static int DumpFrames(AVCodecContext *avctx, davs2_picture_t *pic, davs2_seq_info_t *headerset, AVFrame *frame)
+{
+    DAVS2Context *cad = avctx->priv_data;
+    avctx->flags |= AV_CODEC_FLAG_TRUNCATED;
+
+    if (!headerset)
+        return 0;
+
+    if (!pic || pic->ret_type == DAVS2_GOT_HEADER) {
+        avctx->width        = headerset->horizontal_size;
+        avctx->height       = headerset->vertical_size;
+        avctx->pix_fmt      = (headerset->output_bitdepth == 10 ? AV_PIX_FMT_YUV420P10 : AV_PIX_FMT_YUV420P);
+
+        AVRational r = av_d2q(headerset->frame_rate,4096);
+        avctx->framerate.num = r.num;
+        avctx->framerate.den = r.den;
+        return 0;
+    }
+
+    int bytes_per_sample = pic->bytes_per_sample;
+
+    int i;
+    for (i = 0; i < 3; ++i) {
+        int size_plane = pic->width[i] * pic->lines[i] * bytes_per_sample;
+        frame->buf[i]      = av_buffer_alloc(size_plane);
+        frame->data[i]     = frame->buf[i]->data;
+        frame->linesize[i] = pic->width[i] * bytes_per_sample;
+        if (!frame->buf[i] || !frame->data[i] || !frame->linesize[i]){
+            av_log(avctx, AV_LOG_ERROR, "dump error: alloc failed.\n");
+            return AVERROR(EINVAL);
+        }
+        memcpy(frame->data[i], pic->planes[i], size_plane);
+    }
+
+    frame->width     = cad->headerset.horizontal_size;
+    frame->height    = cad->headerset.vertical_size;
+    frame->pts       = cad->out_frame.pts;
+    frame->pict_type = pic->type;
+    frame->format    = avctx->pix_fmt;
+
+    cad->decoded_frames++;
+    return 1;
+}
+
+static av_cold ff_davs2_end(AVCodecContext *avctx)
+{
+    DAVS2Context *cad = avctx->priv_data;
+
+    /* close the decoder */
+    if (cad->decoder) {
+        davs2_decoder_close(cad->decoder);
+        av_log(avctx, AV_LOG_VERBOSE, "decoder destroyed. 0x%p; frames %d\n", cad->decoder, cad->decoded_frames);
+        cad->decoder = NULL;
+    }
+
+    return 0;
+}
+
+static int davs2_decode_frame(AVCodecContext *avctx, void *data, int *got_frame, AVPacket *avpkt)
+{
+    DAVS2Context *cad = avctx->priv_data;
+    int buf_size       = avpkt->size;
+    uint8_t *buf_ptr = avpkt->data;
+    AVFrame *frame = data;
+    int ret = 0;
+
+    *got_frame = 0;
+    cad->frame = frame;
+    avctx->flags |= AV_CODEC_FLAG_TRUNCATED;
+
+    if (!buf_size) {
+        cad->packet.data = buf_ptr;
```

Comment:
avpkt->data and avpkt->size? We don't need another variable in this case

Reply:
we need this two pointers when we send the next part of data.